odoo/addons/base/models/res_users.py

```diff
@@ -1098,16 +1098,18 @@ def _apply_group(self, implied_group):
         """ Add the given group to the groups implied by the current group
         :param implied_group: the implied group to add
         """
-        if implied_group not in self.implied_ids:
-            self.write({'implied_ids': [Command.link(implied_group.id)]})
+        for group in self:
+            if implied_group not in group.implied_ids:
+                group.write({'implied_ids': [Command.link(implied_group.id)]})
```

Comment:
That sems a lot more inefficient than necessary, the write can still be performed in bulk by selecting all the groups which don't have the `implied_group` and writing them.

Or alternatively by changing the condition to an `any()`/`all()` condition on all groups then writing everything anyway (the latter assumes that groups would mostly not have the new implied group).

Reply:
I pushed a new version using `filtered`. I expect it to be more efficient, at least it emits only one `write`.

I also added a test.